src/test/data/script_tests.json

```diff
@@ -2527,6 +2527,344 @@
 ["0x050000000001", "CHECKSEQUENCEVERIFY", "CHECKSEQUENCEVERIFY", "UNSATISFIED_LOCKTIME",
   "CSV fails if stack top bit 1 << 31 is not set, and tx version < 2"],
 
+["OP_CAT tests"],
+[
+    [
+        "CAT",
```

Comment:
This looks like a violation of the file format. Since these are supposed to be witness stack elements, what should go here are the witnesses as well as the full serialized script.

Maybe an alternative way to do it is to auto generate the serialized script, by defining something like `"<AUTOGEN:SERIALIZED_SCRIPT:'OP_CAT 0xaabb OP_EQUAL'>"`

Reply:
Sorryi f I am misunderstanding you. If I am please rephrase your comment.

`"CAT", control-block` is the full witness stacvk (a.k.a. wit...). This test is checking what happens if you call CAT ("0xe7") on a witness stack that does not contain any pushes just tapscript and control-block. This is what that test looks like without script encoding or autogen. I believe this is a spendable output in Bitcoin today:

```json
["OP_CAT tests"],
[
    [
        "7e",
        "c0d6889cb081036e0faefa3a35157ad71086b123b2b144b649798b494c300a961d",
        0.00000001
    ],
    "",
    "0x51 0x20 0x932a55d505e59555b48961377eb731a1b83efde522f05e493365b9514a43283f",
    "P2SH,WITNESS,TAPROOT",
    "OK",
    "TAPSCRIPT (CAT) Test of OP_CAT flag by calling CAT on an empty stack. This does not error because no TAPSCRIPT_OP_CAT flag is set so CAT is OP_SUCCESS"
],
```

Reply:
Hm, I see what you mean, and I agree that is spendable output. 

TL;DR: What I tried to say I guess is that the witness is not valid without some pre-processing, and that maybe we should add a autogen rule also for serializing the script.

Longer answer:
The confusing part is that, AFAICT, there is no existing test that parses the script depending on whether it is a valid hex value or not (meaning all existing tests are putting the serialized script in the witness).

A better way IMO, to don't have to depend do a special check for whether the `third from the last element in witness stack`  is hex or not, is to define an 'AUTOGEN' rule for parsing the script. That way one could pre-process the witness elements and keep the witness push loop less messy (ignore my broken C++):

```c++
for (i = 0; i < test[pos].size()-1; i++) {
       witness.stack.push_back(ParseWitnessElement(test[pos][i].get_str()), taprootBuilder);
}
...

std::vector<unsigned char> ParseWitnessElement(element string, taprootBuilder *TaprootBuilder) {
      
 if (element.starts_with("<AUTOGEN:CONTROLBLOCK>")) {
       .... build control block...
      return controlblock.begin()
} else if (element.starts_with("<AUTOGEN:SERIALIZED_SCRIPT>")) {
       auto script = element.substr("<AUTOGEN:SERIALIZED_SCRIPT>".length())
        return ToByteVector(ParseScript(script))
} else {
        return ParseHex(element);
}
          
```

Reply:
I implemented this setup for btcd: https://github.com/btcsuite/btcd/pull/2095/commits/bfefe4321ecdff28810946198a7b57fe48d94279 in case you wanna look at that.

Reply:
> TL;DR: What I tried to say I guess is that the witness is not valid without some pre-processing, and that maybe we should add a autogen rule also for serializing the script.

I feel a little like a lawyer saying this so apologies up front. All witness stack values are being preprocessed from hex to byte values. We are just extending this preprocessing to also allow decoding from script format as well as decoding from hex format.

My criteria one when to use AUTOGEN and went not  to use AUTOGEN but an encoding is:
* AUTOGEN is for things that are not deterministic and can't be done by a simple decoding of the bytes in a string. 

> The confusing part is that, AFAICT, there is no existing test that parses the script depending on whether it is a valid hex value or not (meaning all existing tests are putting the serialized script in the witness).

All existing P2WSH tests are putting the hex encoded script in the witness nothing we do here break backwards compatibility with that. All the P2WSH tests are using script encoding for the scripts so this change to enable script encoding of witness scripts brings witness script tests in line with most of the non-witness script tests.

As far as I can tell none of the existing test cases in script_tests.json are using tapscripts.

> A better way IMO, to don't have to depend do a special check for whether the third from the last element in witness stack is hex or not, is to define an 'AUTOGEN' rule for parsing the script. That way one could pre-process the witness elements and keep the witness push loop less messy (ignore my broken C++): 

I don't think a encoding flag system like this is a bad idea. Something like "SCRIPT-ENCODING:..." to make the signal exactly what encoding the script is encoded as. The codebase has already established a precedent of not doing that.

It might we worth talking this through over video. I want to make sure I fully understand your point here.

Reply:
> I feel a little like a lawyer saying this so apologies up front. All witness stack values are being preprocessed from hex to byte values. 

Haha, fair point.

> We are just extending this preprocessing to also allow decoding from script format as well as decoding from hex format.

Yes, but earlier you knew the format when decoding, with this change you have to "guess" or "try". That is why I want some tag to know how to parse each element.

> I like the idea of having encoding flag so that you can do something like "SCRIPT-ENCODING:..." to signal exactly what encoding the script is encoded as. 

Yes, this is exactly what I want. With that I could just go through each witness element one by one, and know exactly how to parse it.

> However for script-tests codebase there is already an established precedent of not doing that. It seems like it be strange to have this just for tapscript test cases and not for everything else.

There is also no precedent for AUTOGEN, so I guess we are already adding new features for tapscripts at this point.

> It might we worth talking this through over video. I want to make sure I fully understand your point here.

Would be happy to jump on a call!

Reply:
> Yes, but earlier you knew the format when decoding, with this change you have to "guess" or "try". That is why I want some tag to know how to parse each element.

I would argue that we know the format it is (hex or script)

> Would be happy to jump on a call!

send an email to ethan.r.heilman@gmail.com and I'll schedule from there